### PR TITLE
HPCC-27692 Provide LogAccess Column Metadata

### PR DESCRIPTION
**Pull request by rpastrana**
- Enhances logaccess logmap config to include col meta
- Updates ALA and Elastic4hpcclogs yaml
- WsLogAccess::GetLogAccessInfo returns col info

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@GordonSmith this is a feature you requested, it provides logaccess column metadata. Please review this sample response and let me know if it works for your purpose:

```<?xml version="1.0" encoding="UTF-8"?>
<GetLogAccessInfoResponse>
	<Columns>
		<Column>
			<Name>LogEntry</Name>
			<LogType>global</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_jobid</Name>
			<LogType>workunits</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
		</Column>
		<Column>
			<Name>Name</Name>
			<LogType>components</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_audience</Name>
			<LogType>audience</LogType>
			<EnumeratedValues>
				<Item>OPR</Item>
				<Item>USR</Item>
				<Item>PRO</Item>
				<Item>ADT</Item>
			</EnumeratedValues>
			<ColumnMode>DEFAULT</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_class</Name>
			<LogType>class</LogType>
			<EnumeratedValues>
				<Item>DIS</Item>
				<Item>ERR</Item>
				<Item>WRN</Item>
				<Item>INF</Item>
				<Item>PRO</Item>
				<Item>MET</Item>
			</EnumeratedValues>
			<ColumnMode>DEFAULT</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_component</Name>
			<LogType>instance</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
		</Column>
		<Column>
			<Name>Computer</Name>
			<LogType>node</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_message</Name>
			<LogType>message</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_sequence</Name>
			<LogType>logid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_procid</Name>
			<LogType>processid</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_threadid</Name>
			<LogType>threadid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
		</Column>
		<Column>
			<Name>hpcc_log_timestamp</Name>
			<LogType>timestamp</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
		</Column>
	</Columns>
	<RemoteLogManagerType>azureloganalyticscurl</RemoteLogManagerType>
	<RemoteLogManagerConnectionString>{Azure Log Analytics workspace ID goes here}</RemoteLogManagerConnectionString>
</GetLogAccessInfoResponse>

File: esp/scm/ws_logaccess.ecm [hunk 1]
```diff
@@ -52,8 +52,33 @@ ESPRequest GetLogAccessInfoRequest
...
+    timestamp("timestamp")
+};
+
+ESPStruct LogColumn
+{
+    string Name;
+    ESPenum LogColumnType LogType;
+    ESParray<string> EnumeratedValues;
+    LogSelectColumnMode ColumnMode;
+};
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply esp/scm/ws_logaccess.ecm:77]: Certainly, it was originally included, but ultimately removed when considering complexity/feature add. 
I believe this was the static mapping it provided:

```
<Column>
			<Name>LogEntry</Name>
			<LogType>global</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
                         **<ColumnType>String</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_jobid</Name>
			<LogType>workunits</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
                         **<ColumnType>String</ColumnType>**
		</Column>
		<Column>
			<Name>Name</Name>
			<LogType>components</LogType>
			<EnumeratedValues/>
                         **<ColumnType>String</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_audience</Name>
			<LogType>audience</LogType>
			<EnumeratedValues>
				<Item>OPR</Item>
				<Item>USR</Item>
				<Item>PRO</Item>
				<Item>ADT</Item>
			</EnumeratedValues>
			<ColumnMode>DEFAULT</ColumnMode>
                         **<ColumnType>Enum</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_class</Name>
			<LogType>class</LogType>
			<EnumeratedValues>
				<Item>DIS</Item>
				<Item>ERR</Item>
				<Item>WRN</Item>
				<Item>INF</Item>
				<Item>PRO</Item>
				<Item>MET</Item>
			</EnumeratedValues>
			<ColumnMode>DEFAULT</ColumnMode>
                         **<ColumnType>Enum</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_component</Name>
			<LogType>instance</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
                         **<ColumnType>String</ColumnType>**
		</Column>
		<Column>
			<Name>Computer</Name>
			<LogType>node</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
                         **<ColumnType>String</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_message</Name>
			<LogType>message</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
                         **<ColumnType>String</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_sequence</Name>
			<LogType>logid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
                         **<ColumnType>Numeric</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_procid</Name>
			<LogType>processid</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
                         **<ColumnType>Numeric</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_threadid</Name>
			<LogType>threadid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
                         **<ColumnType>Numeric</ColumnType>**
		</Column>
		<Column>
			<Name>hpcc_log_timestamp</Name>
			<LogType>timestamp</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
                         **<ColumnType>DateTime</ColumnType>**
		</Column>



Let me know how important this info is for the UI layer. Thanks. @GordonSmith

[the remark this answers is not in the document]
Comment (GordonSmith) [reply esp/scm/ws_logaccess.ecm:77]: Yea looks good - will be wanted, but not top of our list yet.

**Comment by GordonSmith** [comment]
@rpastrana looks fine, will need to know a list of the possible types, but that doesn't need to be discoverable as we need to handle them differently in code.

File: esp/scm/ws_logaccess.ecm (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpastrana) [reply esp/scm/ws_logaccess.ecm:86]: @GordonSmith I've added the columndatatype field. you can expect this type of output:
```
<?xml version="1.0" encoding="UTF-8"?>
<GetLogAccessInfoResponse>
	<Columns>
		<Column>
			<Name>ixmessage</Name>
			<LogType>global</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
			<ColumnType>string</ColumnType>
		</Column>
		<Column>
			<Name>ixhpcc.log.jobid</Name>
			<LogType>workunits</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
			<ColumnType>string</ColumnType>
		</Column>
		<Column>
			<Name>kubernetes.container.name</Name>
			<LogType>components</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
			<ColumnType>string</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.audience</Name>
			<LogType>audience</LogType>
			<EnumeratedValues>
				<Item>OPR</Item>
				<Item>USR</Item>
				<Item>PRO</Item>
				<Item>ADT</Item>
			</EnumeratedValues>
			<ColumnMode>DEFAULT</ColumnMode>
			<ColumnType>enum</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.class</Name>
			<LogType>class</LogType>
			<EnumeratedValues>
				<Item>DIS</Item>
				<Item>ERR</Item>
				<Item>WRN</Item>
				<Item>INF</Item>
				<Item>PRO</Item>
				<Item>MET</Item>
			</EnumeratedValues>
			<ColumnMode>DEFAULT</ColumnMode>
			<ColumnType>enum</ColumnType>
		</Column>
		<Column>
			<Name>kubernetes.pod.name</Name>
			<LogType>instance</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
			<ColumnType>string</ColumnType>
		</Column>
		<Column>
			<Name>kubernetes.node.hostname</Name>
			<LogType>node</LogType>
			<EnumeratedValues/>
			<ColumnMode>ALL</ColumnMode>
			<ColumnType>string</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.message</Name>
			<LogType>message</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
			<ColumnType>string</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.sequence</Name>
			<LogType>logid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
			<ColumnType>numeric</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.procid</Name>
			<LogType>processid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
			<ColumnType>numeric</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.threadid</Name>
			<LogType>threadid</LogType>
			<EnumeratedValues/>
			<ColumnMode>DEFAULT</ColumnMode>
			<ColumnType>numeric</ColumnType>
		</Column>
		<Column>
			<Name>hpcc.log.timestamp</Name>
			<LogType>timestamp</LogType>
			<EnumeratedValues/>
			<ColumnMode>MIN</ColumnMode>
			<ColumnType>datetime</ColumnType>
		</Column>
	</Columns>
	<RemoteLogManagerType>elasticstack</RemoteLogManagerType>
	<RemoteLogManagerConnectionString>http://elasticsearch-master.default.svc.cluster.local:9200/</RemoteLogManagerConnectionString>
</GetLogAccessInfoResponse>
```

**Comment by rpastrana** [comment]
@wangkx please review

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.hpp [hunk 2]
```diff
@@ -82,8 +81,12 @@ class AzureLogAnalyticsCurlClient : public CInterfaceOf<IRemoteLogAccess>
     // IRemoteLogAccess methods
     virtual bool fetchLog(LogQueryResultDetails & resultDetails, const LogAccessConditions & options, StringBuffer & returnbuf, LogAccessLogFormat format) override;
     virtual const char * getRemoteLogAccessType() const override { return type; }
-    virtual IPropertyTree * queryLogMap() const override { return m_pluginCfg->queryPropTree("logmap");}
-    virtual const char * fetchConnectionStr() const override { return m_logAnalyticsWorkspaceID.str();}
+    virtual IPropertyTree * queryLogMap() const override
+    {
+        DBGLOG("AzureLogAnalyticsCurlClient::queryLogMap");
```
Comment (wangkx) [inline system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.hpp:86]: Do you still need this debug code line?

File: esp/services/ws_logaccess/WsLogAccessService.cpp [hunk 3]
```diff
@@ -13,11 +13,85 @@ bool Cws_logaccessEx::onGetLogAccessInfo(IEspContext &context, IEspGetLogAccessI
...
+                IArrayOf<IEspLogColumn> logColumns;
+                Owned<IPropertyTreeIterator> columnIter = logColumnDefinitions->getElements("logMaps");
+                ForEach(*columnIter)
+                {
+                    IPropertyTree & column = columnIter->query();
+                    if (column.hasProp("@searchColumn"))
+                    {
+                        Owned<IEspLogColumn> espLogColumn = createLogColumn();
+
+                        const char * csearchColumn=column.queryProp("@searchColumn");
```
Comment (wangkx) [inline esp/services/ws_logaccess/WsLogAccessService.cpp:37]: trivial: spaces in both sides of the '='.

File: esp/services/ws_logaccess/WsLogAccessService.cpp [hunk 4]
```diff
@@ -13,11 +13,85 @@ bool Cws_logaccessEx::onGetLogAccessInfo(IEspContext &context, IEspGetLogAccessI
...
+                        const char * cprojectName=column.queryProp("@projectName");
```
Comment (wangkx) [inline esp/services/ws_logaccess/WsLogAccessService.cpp:38]: trivial: spaces in both sides of the '='.

File: esp/services/ws_logaccess/WsLogAccessService.cpp [hunk 5]
```diff
@@ -13,11 +13,85 @@ bool Cws_logaccessEx::onGetLogAccessInfo(IEspContext &context, IEspGetLogAccessI
     bool success = true;
     if (queryRemoteLogAccessor())
     {
+        IPropertyTree * logColumnDefinitions = queryRemoteLogAccessor()->queryLogMap();
```
Comment (wangkx) [inline esp/services/ws_logaccess/WsLogAccessService.cpp:16]: trivial: to be consistent with line 11, remove the space from the front of the '*'.

File: esp/services/ws_logaccess/WsLogAccessService.cpp [hunk 6]
```diff
@@ -13,11 +13,85 @@ bool Cws_logaccessEx::onGetLogAccessInfo(IEspContext &context, IEspGetLogAccessI
...
+        {
+            double clientVer = context.getClientVersion();
+
+            if (clientVer >= 1.04)
+            {
+                IArrayOf<IEspLogColumn> logColumns;
+                Owned<IPropertyTreeIterator> columnIter = logColumnDefinitions->getElements("logMaps");
+                ForEach(*columnIter)
+                {
+                    IPropertyTree & column = columnIter->query();
```
Comment (wangkx) [inline esp/services/ws_logaccess/WsLogAccessService.cpp:32]: trivial: to be consistent with line 11, remove the space from the front of the '*'.

File: esp/services/ws_logaccess/WsLogAccessService.cpp [hunk 7]
```diff
@@ -13,11 +13,85 @@ bool Cws_logaccessEx::onGetLogAccessInfo(IEspContext &context, IEspGetLogAccessI
...
+                        if (column.hasProp("@columnMode"))
+                            espLogColumn->setColumnMode(column.queryProp("@columnMode"));
+
+                        if (column.hasProp("EnumValues"))
+                        {
+                            Owned<IPropertyTreeIterator> enumValues = column.getElements("EnumValues");
+                            StringArray enumValuesArr;
+                            ForEach(*enumValues)
+                            {
+                                IPropertyTree & enumVal = enumValues->query();
```
Comment (wangkx) [inline esp/services/ws_logaccess/WsLogAccessService.cpp:53]: trivial: to be consistent with line 11, remove the space from the front of the '*'.

**Review by wangkx — COMMENTED** [review]
@rpastrana one question and a few minor style comments

**Comment by rpastrana** [comment]
@wangkx please see code review changes

**Review by wangkx — APPROVED** [review]
@rpastrana looks good.

**Comment by rpastrana** [comment]
@GordonSmith unless you can spot any issues, we'll ask to merge this PR. thanks.

**Comment by rpastrana** [comment]
@GordonSmith are we good w/ this?

**Comment by ghalliday** [comment]
Waiting for input from Gordon, but I will tag the next 8.10.x without it to minimize the chance of any problems since it contains a fix for a critical regression.

**Review by GordonSmith — APPROVED** [review]

**Comment by rpastrana** [comment]
@ghalliday ready for merge.

**Comment by rpastrana** [comment]
@ghalliday this is ready to merge, should it be retargeted to 8.12.x?

**Review by ghalliday — COMMENTED** [review]
@rpastrana apologies for the delay in reviewing this PR.  A few fairly minor comments.

File: esp/services/ws_logaccess/WsLogAccessService.cpp [hunk 8]
```diff
@@ -13,11 +13,85 @@ bool Cws_logaccessEx::onGetLogAccessInfo(IEspContext &context, IEspGetLogAccessI
...
+
+                        if (column.hasProp("@type"))
+                            espLogColumn->setLogType(column.queryProp("@type"));
+
+                        if (column.hasProp("@columnMode"))
+                            espLogColumn->setColumnMode(column.queryProp("@columnMode"));
+
+                        if (column.hasProp("EnumValues"))
+                        {
+                            Owned<IPropertyTreeIterator> enumValues = column.getElements("EnumValues");
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply esp/services/ws_logaccess/WsLogAccessService.cpp:49]: Spent some time looking for this normal standard (would have been trivial if we had it written up, and referenced directly), anyhow, all ECM fields appear to be capitalized and this follows that standard, are you suggesting we break that standard?

File: helm/hpcc/values.schema.json [hunk 9]
```diff
@@ -2756,7 +2756,7 @@
         "type": {
           "type": "string",
           "description" : "The searchable HPCC log column to be mapped - 'global' applies to all known fields",
-          "enum": [ "global", "workunits", "components", "audience", "class", "instance", "host"]
+          "enum": [ "global", "workunits", "components", "audience", "class", "instance", "node", "message", "logid", "processid", "threadid", "timestamp"]
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply helm/hpcc/values.schema.json:2759]: good question. As of now, eclwatch is the only consumer of this service, and it does not appear to be an issue for eclwatch.

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.hpp [hunk 10]
```diff
@@ -82,8 +81,8 @@ class AzureLogAnalyticsCurlClient : public CInterfaceOf<IRemoteLogAccess>
     // IRemoteLogAccess methods
     virtual bool fetchLog(LogQueryResultDetails & resultDetails, const LogAccessConditions & options, StringBuffer & returnbuf, LogAccessLogFormat format) override;
     virtual const char * getRemoteLogAccessType() const override { return type; }
-    virtual IPropertyTree * queryLogMap() const override { return m_pluginCfg->queryPropTree("logmap");}
-    virtual const char * fetchConnectionStr() const override { return m_logAnalyticsWorkspaceID.str();}
+    virtual IPropertyTree * queryLogMap() const override { return m_pluginCfg->queryPropTree(""); }
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.hpp:84]: This change is needed for the way the logcolumndefinitions are queried at the ESP service:

IPropertyTree *logColumnDefinitions = queryRemoteLogAccessor()->queryLogMap();

File: helm/examples/azure/log-analytics/loganalytics-hpcc-logaccess.yaml [hunk 11]
```diff
@@ -19,23 +19,80 @@ global:
...
       searchColumn: "Name"
       keyColumn: "ContainerID"
-      timeStampColumn: "TimeGenerated"
+      columnMode: "MIN"
+      columnType: "string"
     - type: "audience"
       searchColumn: "hpcc_log_audience"
+      EnumValues:
+      - Code: OPR
+        Label: Operator
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply helm/examples/azure/log-analytics/loganalytics-hpcc-logaccess.yaml:39]: yes, this was intended to for the UI layer, but I did not push it through to the ESP response.
@GordonSmith how do you feel about the service providing the extended name for the enumerated values?

**Comment by GordonSmith** [comment]
@ghalliday - myself and/or @jeclrsg are going to build and test this as part of the review process.

**Comment by GordonSmith** [comment]
@rpastrana - should this be targeting master?

**Comment by rpastrana** [comment]
> @rpastrana - should this be targeting master?

yep, as discussed this is a 9.0.0 feature. Now targeting master

**Comment by rpastrana** [comment]
Looked over errors in checks, but they do not seem related to the pr

**Comment by rpastrana** [comment]
@ghalliday please review latest change

**Review by ghalliday — APPROVED** [review]

**Review by ghalliday — APPROVED** [review]

**Review by ghalliday — APPROVED** [review]
approved

**Review by GordonSmith — APPROVED** [review]
Looks good - but may need to open new issues when we start integration...

**Comment by rpastrana** [comment]
@ghalliday please merge for 9.0.0

**Pull request merged**